### PR TITLE
[feat] add set as a data structure

### PR DESCRIPTION
**Pull request by kopancek**
## Description

This is a thing I was long dreaming about, since our current code for dealing with permission sets is clunky to say the least.

## Test plan

Unit tests, it's not used anywhere yet, but plan on refactoring the perms_store as a next step.

### CODE REVIEW

File: internal/collections/set.go [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+// or with NewSet[T]().
+type Set[T comparable] map[T]struct{}
+
+// NewSet creates a Set[T] with the given values.
+// T must be a comparable type (implementing sort.Interface or == operator).
+//
+// Example:
+//
+//	s := NewSet[int](1, 2, 3)
+func NewSet[T comparable](values ...T) Set[T] {
```
Comment (sashaostrikov) [inline internal/collections/set.go:21]: Should we add `String()` as well for logging purposes?

File: internal/collections/set.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+// SortedValues returns the values of the set in sorted order using the given
+// comparator function.
+//
+// The comparator function should return true if the first argument is less than
+// the second, and false otherwise.
+//
+// Example:
+//
+//	s.SortedValues(func(a, b int) bool { return a < b })
+func (s Set[T]) SortedValues(comparator func(a, b T) bool) []T {
```
Comment (sashaostrikov) [inline internal/collections/set.go:57]: nit: might as well be called `Sorted`

**Review by sashaostrikov — APPROVED** [review]
LGTM!

File: internal/collections/set.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+}
+
+func (s Set[T]) Remove(value T) {
+	delete(s, value)
+}
+
+func (s Set[T]) Contains(value T) bool {
+	_, found := s[value]
+	return found
+}
```
[the remark this answers is not in the document]
Comment (kopancek) [reply internal/collections/set.go:41]: I like the Add and Remove to allow multiple values, however contains does not really make sense to me. I'll add another method `IntersectsWith` that will spit out a bool if one set is within the other one.

**Review by mrnugget — APPROVED** [review]
good stuff! wanna see where we use it :) previously I haven't had a lot of success with a `Set` in our codebase, becuase it'd introduce double loops, but let's see

File: internal/collections/set.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (kopancek) [reply internal/collections/set.go]: In the end I renamed the existing Contains to `Has` and `Contains` now checks if one set contains another set.

**Pull request merged**